File: apps/web/src/lib/timeRange.ts
// Shared time-range filter model: preset windows plus a custom range (e.g. set by
// drag-selecting on a chart). Bounds are UTC-normalized to match the app's UTC day grouping.

export type TimeRangePreset = '7d' | '30d' | '90d' | 'ytd' | 'all';

export type TimeRangeValue = { preset: TimeRangePreset } | { preset: 'custom'; start: Date; end: Date };

export interface ResolvedTimeRange {
    start: Date | null;
    end: Date | null;
}

const startOfUTCDay = (d: Date): Date => {
    const out = new Date(d);
    out.setUTCHours(0, 0, 0, 0);
    return out;
};

const endOfUTCDay = (d: Date): Date => {
    const out = new Date(d);
    out.setUTCHours(23, 59, 59, 999);
    return out;
};

/** UTC midnight `days - 1` days ago, so the window includes today. */
const rollingWindowStart = (days: number): Date => {
    const d = startOfUTCDay(new Date());
    d.setUTCDate(d.getUTCDate() - days + 1);
    return d;
};

/** Null bounds mean unbounded on that side. */
export function resolveTimeRange(value: TimeRangeValue): ResolvedTimeRange {
    switch (value.preset) {
        case '7d':
            return { start: rollingWindowStart(7), end: null };
        case '30d':
            return { start: rollingWindowStart(30), end: null };
        case '90d':
            return { start: rollingWindowStart(90), end: null };
        case 'ytd':
            return { start: new Date(Date.UTC(new Date().getUTCFullYear(), 0, 1)), end: null };
        case 'all':
            return { start: null, end: null };
        case 'custom':
            return { start: startOfUTCDay(value.start), end: endOfUTCDay(value.end) };
    }
}

export function inTimeRange(date: Date, range: ResolvedTimeRange): boolean {
    if (range.start && date < range.start) return false;
    if (range.end && date > range.end) return false;
    return true;
}

File: apps/native/Sources/Models/TimeRange.swift
import Foundation

/// The window the exercise page plots over. Hand-mirrored from `apps/web/src/lib/timeRange.ts`,
/// minus its `custom` case — on the web a chart drag selects a range, here a drag scrubs.
enum TimeRange: String, CaseIterable, Identifiable, Hashable {
    case sevenDays = "7D"
    case thirtyDays = "30D"
    case yearToDate = "YTD"
    case all = "MAX"

    var id: String { rawValue }

    /// Every date in this app is a UTC day key, so the window boundaries are too. A device
    /// calendar would move them a day west of Greenwich — the same off-by-one CLAUDE.md warns
    /// about for `DatePicker`.
    static let utc: Calendar = {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC")!
        calendar.locale = Locale(identifier: "en_US_POSIX")
        return calendar
    }()

    var resolved: ResolvedTimeRange {
        switch self {
        case .sevenDays: ResolvedTimeRange(start: Self.rollingStart(days: 7), end: nil)
        case .thirtyDays: ResolvedTimeRange(start: Self.rollingStart(days: 30), end: nil)
        case .yearToDate:
            ResolvedTimeRange(
                start: Self.utc.date(from: DateComponents(year: Self.utc.component(.year, from: .now), month: 1, day: 1)),
                end: nil
            )
        case .all: ResolvedTimeRange(start: nil, end: nil)
        }
    }

    /// `days - 1` back from UTC midnight today, so the window includes today.
    private static func rollingStart(days: Int) -> Date? {
        utc.date(byAdding: .day, value: -(days - 1), to: utc.startOfDay(for: .now))
    }
}

/// A resolved window. A nil bound is unbounded on that side, so `all` is nil on both — which is
/// also how the star rule recognises an unfiltered view.
struct ResolvedTimeRange {
    let start: Date?
    let end: Date?

    var isUnbounded: Bool { start == nil && end == nil }

    func contains(_ date: Date) -> Bool {
        if let start, date < start { return false }
        if let end, date > end { return false }
        return true
    }
}
